SmartDeviceLink/SDLMenuManager.m

```diff
@@ -0,0 +1,400 @@
+        }
+    }
+
+    return [mutableDeletes copy];
+}
+
+#pragma mark Commands / SubMenu RPCs
+
+- (NSArray<SDLRPCRequest *> *)sdl_mainMenuCommandsForCells:(NSArray<SDLMenuCell *> *)cells withArtwork:(BOOL)shouldHaveArtwork {
+    NSMutableArray<SDLRPCRequest *> *mutableCommands = [NSMutableArray array];
+    [cells enumerateObjectsUsingBlock:^(SDLMenuCell * _Nonnull cell, NSUInteger index, BOOL * _Nonnull stop) {
+        if (cell.subCells.count > 0) {
+            [mutableCommands addObject:[self sdl_subMenuCommandForMenuCell:cell position:(UInt16)index]];
+        } else {
+            [mutableCommands addObject:[self sdl_commandForMenuCell:cell withArtwork:shouldHaveArtwork position:(UInt16)index]];
+        }
+    }];
+
+    return [mutableCommands copy];
+}
+
+- (NSArray<SDLRPCRequest *> *)sdl_subMenuCommandsForCells:(NSArray<SDLMenuCell *> *)cells withArtwork:(BOOL)shouldHaveArtwork {
+    NSMutableArray<SDLRPCRequest *> *mutableCommands = [NSMutableArray array];
+    for (SDLMenuCell *cell in cells) {
+        if (cell.subCells.count > 0) {
+            [mutableCommands addObjectsFromArray:[self sdl_allCommandsForCells:cell.subCells withArtwork:shouldHaveArtwork]];
+        }
+    }
+
+    return [mutableCommands copy];
+}
+
+- (NSArray<SDLRPCRequest *> *)sdl_allCommandsForCells:(NSArray<SDLMenuCell *> *)cells withArtwork:(BOOL)shouldHaveArtwork {
+    NSMutableArray<SDLRPCRequest *> *mutableCommands = [NSMutableArray array];
+    [cells enumerateObjectsUsingBlock:^(SDLMenuCell * _Nonnull cell, NSUInteger index, BOOL * _Nonnull stop) {
+        if (cell.subCells.count > 0) {
+            [mutableCommands addObject:[self sdl_subMenuCommandForMenuCell:cell position:(UInt16)index]];
+            [mutableCommands addObjectsFromArray:[self sdl_allCommandsForCells:cell.subCells withArtwork:shouldHaveArtwork]];
+        } else {
+            [mutableCommands addObject:[self sdl_commandForMenuCell:cell withArtwork:shouldHaveArtwork position:(UInt16)index]];
+        }
+    }];
+
+    return [mutableCommands copy];
+}
+
+- (SDLAddCommand *)sdl_commandForMenuCell:(SDLMenuCell *)cell withArtwork:(BOOL)shouldHaveArtwork position:(UInt16)position {
+    SDLAddCommand *command = [[SDLAddCommand alloc] init];
+
+    SDLMenuParams *params = [[SDLMenuParams alloc] init];
+    params.menuName = cell.title;
+    params.parentID = cell.parentCellId != UINT32_MAX ? @(cell.parentCellId) : nil;
+    params.position = @(position);
+
+    command.menuParams = params;
+    command.vrCommands = cell.voiceCommands;
+    command.cmdIcon = (cell.icon && shouldHaveArtwork) ? [[SDLImage alloc] initWithName:cell.icon.name] : nil;
+    command.cmdID = @(cell.cellId);
+
+    return command;
+}
+
+- (SDLAddSubMenu *)sdl_subMenuCommandForMenuCell:(SDLMenuCell *)cell position:(UInt16)position {
+    SDLAddSubMenu *submenu = [[SDLAddSubMenu alloc] initWithId:cell.cellId menuName:cell.title];
+    submenu.position = @(position);
+
```

Comment:
If an `AddSubMenu` is the first menu item, I get an `INVALID_DATA` response. It seems to work fine if the `AddSubMenu` is the second menu item. Not sure why?

```
2018-04-25 11:33:23.633037-0400 SDL_Smoke_Tests[1112:81862] [Dispatcher] Request failed: AddSubMenu (request)
{
    menuID = 0;
    menuName = "Root 1";
    position = 0;
}, response: AddSubMenu (response)
{
    resultCode = "INVALID_DATA";
    success = 0;
}, error: Error Domain=com.sdl.lifecyclemanager.error Code=-1 "INVALID_DATA" UserInfo={NSLocalizedRecoverySuggestion=Have you tried turning it off and on again?, NSLocalizedDescription=INVALID_DATA, NSLocalizedFailureReason=}
```

Reply:
Seems to be a Core bug?

Reply:
Fixed